### PR TITLE
Response withExpiredCookie

### PR DESCRIPTION
**Pull request by jorisvaesen**
Refs #10975

### CODE REVIEW

File: src/Http/Response.php [hunk 1]
```diff
@@ -2041,6 +2042,66 @@ public function withCookie($name, $data = '')
...
+     * $response = $response->withoutCookie('remember_me');
+     *
+     * // customize cookie attributes
+     * $response = $response->withoutCookie('remember_me', ['path' => '/login']);
+     *
+     * // add a cookie object
+     * $response = $response->withoutCookie(new Cookie('remember_me', 'deleted'));
+     * ```
+     *
+     * @param string|\Cake\Http\Cookie\Cookie $name The name of the cookie to expire, or a cookie object
```
Comment (robertpustulka) [inline src/Http/Response.php]: Should be `string|\Cake\Http\Cookie\CookieInterface`.

File: src/Http/Response.php [hunk 2]
```diff
@@ -2041,6 +2042,66 @@ public function withCookie($name, $data = '')
...
+     * @param array $options An array of cookie options.
+     * @return static
+     */
+    public function withoutCookie($name, $options = [])
+    {
+        if ($name instanceof Cookie) {
```
Comment (robertpustulka) [inline src/Http/Response.php]: `$name instanceof CookieInterface`

File: src/Http/Response.php [hunk 3]
```diff
@@ -2041,6 +2042,66 @@ public function withCookie($name, $data = '')
...
+     *
+     * // add a cookie object
+     * $response = $response->withoutCookie(new Cookie('remember_me', 'deleted'));
+     * ```
+     *
+     * @param string|\Cake\Http\Cookie\Cookie $name The name of the cookie to expire, or a cookie object
+     * @param array $options An array of cookie options.
+     * @return static
+     */
+    public function withoutCookie($name, $options = [])
```
Comment (robertpustulka) [inline src/Http/Response.php]: `array` type hint would be nice.

Comment (markstory) [inline src/Http/Response.php]: Is this method 'removing' a cookie from the response that is in flight or adding an expired cookie? Right now `withoutCookie()` will add a cookie which is a bit counter-intuitive.

File: src/Http/Response.php [hunk 4]
```diff
@@ -2041,6 +2042,66 @@ public function withCookie($name, $data = '')
...
+        if ($name instanceof Cookie) {
+            $cookie = $name->withExpiry(new Time(1));
+        } else {
+            $options += [
+                'value' => '',
+                'path' => '/',
+                'domain' => '',
+                'secure' => false,
+                'httpOnly' => false
+            ];
```
Comment (markstory) [inline src/Http/Response.php]: If the cookie exists in the collection, would it make sense to read the options out of that cookie object?

File: tests/TestCase/Http/ResponseTest.php [hunk 5]
```diff
@@ -1491,6 +1491,53 @@ public function testWithCookieObject()
...
+            'secure' => true,
+            'httpOnly' => true,
+            'expire' => (string)strtotime('+14 days'),
+        ];
+
+        $response = new Response();
+        $response = $response->withCookie('testing', $options);
+        $this->assertEquals($options, $response->getCookie('testing'));
+
+        $new = $response->withoutCookie('testing', $options);
```
Comment (markstory) [inline tests/TestCase/Http/ResponseTest.php]: expire as an option to this method should probably trigger an error, as should 'value'.  At the very least those options should be ignored. The test doesn't cover the value being ignored.

File: src/Http/Response.php [hunk 6]
```diff
@@ -2041,6 +2042,66 @@ public function withCookie($name, $data = '')
     }
 
     /**
+     * Create a new response with an expired cookie set.
+     *
+     * ### Options
+     *
+     * - `name`: The Cookie name
+     * - `value`: Value of the cookie
```
Comment (markstory) [inline src/Http/Response.php]: Value shouldn't be an option. We can hardcode the value to ''.

File: src/Http/Response.php (hunk 3 above)
Comment (jorisvaesen) [inline src/Http/Response.php]: @markstory Removing means calling `remove` on the CookieCollection and reapply it.
Nonetheless, I agree that this naming isn't very convenient but the CookieComponent is working the same where `delete` has the same functionality. How about `withExpiredCookie()`?

Comment (markstory) [inline src/Http/Response.php]: I like `withExpiredCookie()`. It better describes what this method is doing.

File: src/Http/Response.php [hunk 7]
```diff
@@ -19,6 +19,7 @@
 use Cake\Http\Cookie\Cookie;
 use Cake\Http\Cookie\CookieCollection;
 use Cake\Http\Cookie\CookieInterface;
+use Cake\I18n\Time;
```
Comment (robertpustulka) [inline src/Http/Response.php]: I'm wondering if there's a need to depend on `I18n`?

File: src/Http/Cookie/Cookie.php [hunk 8]
```diff
@@ -510,7 +510,7 @@ public function withNeverExpire()
     public function withExpired()
     {
         $new = clone $this;
-        $new->expiresAt = Chronos::parse('-1 year');
+        $new->expiresAt = Chronos::createFromTimestamp(1);
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php]: Why? -1 year is good enough to expire cookies.

File: src/Http/Response.php [hunk 9]
```diff
@@ -2058,26 +2056,25 @@ public function withCookie($name, $data = '')
...
      * @param array $options An array of cookie options.
      * @return static
      */
-    public function withoutCookie($name, $options = [])
+    public function withExpiredCookie($name, $options = [])
     {
-        if ($name instanceof Cookie) {
-            $cookie = $name->withExpiry(new Time(1));
+        if ($name instanceof CookieInterface) {
+            $cookie = $name->withExpired();
```
Comment (markstory) [inline src/Http/Response.php]: :+1:

File: src/Http/Response.php [hunk 10]
```diff
@@ -2058,26 +2056,25 @@ public function withCookie($name, $data = '')
...
         } else {
             $options += [
-                'value' => '',
                 'path' => '/',
```
Comment (markstory) [inline src/Http/Response.php]: Should this default to `$this->getAttribute('webroot')`?

File: src/Http/Cookie/Cookie.php (hunk 8 above)
Comment (jorisvaesen) [inline src/Http/Cookie/Cookie.php]: For me, it's not a worse option than `-1 year` (and looks like the safest and most consistent) and is now the same as how the CAKEPHP-cookie gets expired.

File: src/Http/Response.php (hunk 10 above)
Comment (markstory) [inline src/Http/Response.php]: Ignore this. My suggestion is not possible.

**Pull request merged**